File: FishTrackerLambda/src/FishTrackerLambda.Services/Services/TripService.cs
using Amazon.DynamoDBv2;
using FishTrackerLambda.DataAccess;
using FishTrackerLambda.Functional;
using FishTrackerLambda.Helpers;
using FishTrackerLambda.Models.Lambda;
using FishTrackerLambda.Models.Persistance;
using FishTrackerLambda.Services.Http;
using Microsoft.Extensions.Logging;

namespace FishTrackerLambda.Services
{
    public class TripService : ITripService
    {
        private readonly ILogger<TripService> m_logger;

        private readonly IAmazonDynamoDB m_client;

        public TripService(ILogger<TripService> logger, IAmazonDynamoDB client)
        {
            m_logger = logger;
            m_client = client;
        }

        public Task<HttpWrapper<TripDetails>> GetTrip(string subject, string tripId)
        {
            return TripDbTable
                .ReadTripFromDynamodb(subject, tripId, m_client, m_logger)
                .Map(c => c.ToTripDetails());
        }

        public Task<HttpWrapper<IEnumerable<TripDetails>>> GetTrips(string subject, string? view)
        {
            switch (view)
            {
            case null:
            case "all":
                return Function
                    .InitAsync(TripDbTable.ReadAllTripsFromDynamoDb(subject, m_client, m_logger))
                    .Map(c => c.Select(r => r.ToTripDetails()));
            case "relevant":
                return Function
                    .InitAsync(TripDbTable.ReadRelevantTripsFromDynamoDb(subject, m_client, m_logger))
                    .Map(c => c.Select(r => r.ToTripDetails()));
            default:
                throw new Exception($"invalid view[{view}]");
            }
        }


        public Task<HttpWrapper<TripDetails>> NewTrip(string subject, NewTrip newTrip)
        {
            return Function
                .ValidateInput(() => {
                    return newTrip.startTime != null || newTrip.timeZone != null ? null : Results.BadRequest("Must supply either a datetime or timezone");
                 }) 
                .Init(newTrip.FillInMissingData())
                .Map(t => t.CreateNewDyanmoRecord(subject))
                .MapAsync(t => t.WriteTripToDynamoDb(m_client, m_logger))
                .Map(t => t.ToTripDetails());
        }

        public Task<HttpWrapper<TripDetails>> UpdateTrip(string subject, string tripId, TripDetails trip)
        {
            return Function
                .ValidateInput(() => {
                    return tripId == trip.tripId ? null : Results.BadRequest($"Cannot change tripId from[{tripId}] to[{trip.tripId}]");
                 })
                .MapAsync(t => TripDbTable.ReadTripFromDynamodb(subject, trip.tripId, m_client, m_logger))
                .Map(t => t.UpdateTrip(trip))
                .MapAsync(t => t.UpdateTripInDynamodb(m_client, m_logger))
                .Map(t => t.ToTripDetails());
        }

        public Task<HttpWrapper<TripDetails>> PatchTrip(string subject, string tripId, UpdateTripDetails trip)
        {
            return TripDbTable
                .ReadTripFromDynamodb(subject, tripId, m_client, m_logger)
                .Map(c => c.PatchTrip(trip))
                .MapAsync(c => c.UpdateTripInDynamodb(m_client, m_logger))
                .Map(c => c.ToTripDetails());
        }

        public Task<HttpWrapper<IEnumerable<CatchDetails>>> DeleteTrip(string subject, string tripId)
        {
            return TripDbTable
                .ReadTripFromDynamodb(subject, tripId, m_client, m_logger)
                .MapAsync(c => c.DeleteTripInDynamodb(m_client, m_logger))
                .MapAsync(c => c.ReadAllCatchFromDynamoDb(m_client, m_logger))
                .MapEachAsync(c => c.DeleteCatchInDynamodb(m_client, m_logger))
                .Map(c => c.Select(r => r.ToCatchDetails()));
        }

        public Task<HttpWrapper<TripDetails>> EndTrip(string subject, string tripId, EndTripDetails trip)
        {
            return CatchDbTable.ReadAllCatchFromDynamoDb(IdGenerator.GenerateTripKey(subject, tripId), m_client, m_logger)
                .MapAsync(all =>
                {
                    return TripDbTable
                        .ReadTripFromDynamodb(subject, tripId, m_client, m_logger)
                        .Map(c => c.EndTrip(trip, all.Count()))
                        .MapAsync(c => c.UpdateTripInDynamodb(m_client, m_logger))
                        .Map(c => c.ToTripDetails());
                });

        }
    }
}

File: FishTrackerLambdaNode/src/Services/TripService.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { container, inject, injectable } from 'tsyringe';
import { ITripDetails, INewTrip, IUpdateTripDetails, IEndTripDetails, IDynamoDbCatch, ICatchDetails } from '../Models/lambda';
import { TripDbService } from '../Db.Services/TripDbService';
import { CatchDbService } from '../Db.Services/CatchDbService';
import { HttpWrapper } from '../Functional/HttpWrapper';
import { IdGenerator } from '../Helpers/IdGenerator';
import { Logger } from '@aws-lambda-powertools/logger';
import { Results } from '../Http/Result';

@injectable()
export class TripService {
    private tripService: TripDbService;
    private catchService: CatchDbService;
    private logger: Logger;

    constructor(tripService: TripDbService, catchService: CatchDbService, logger: Logger) {
        this.tripService = tripService
        this.catchService = catchService
        this.logger = logger
    }

    public async getTrip(subject: string, tripId: string): Promise<HttpWrapper<ITripDetails>> {
        const tripRecord = await this.tripService.readRecordWithSortKey(subject, tripId);
        return tripRecord.Map(c => TripDbService.toTripDetails(c));
    }

    public async getTrips(subject: string, view?: string): Promise<HttpWrapper<ITripDetails[]>> {
        switch (view) {
            case null:
            case 'all':
                return (await this.tripService
                    .readAllRecordsForPartition(subject))
                    .Map(c => {console.log(c); return c})
                    .Map(c => {console.log(JSON.stringify(c)); return c})
                    .Map(c => {this.logger.info(JSON.stringify(c)); return c})
                    .Map(c => c.map(r => TripDbService.toTripDetails(r)))
                    .Map(c => {this.logger.info(JSON.stringify(c)); return c})
            case 'relevant':
                return (await this.tripService
                    .readRelevantTripsFromDynamoDb(subject))
                    .Map(c => c.map(r => TripDbService.toTripDetails(r)));
            default:
                throw new Error(`invalid view[${view}]`);
        }
    }

    public async newTrip(subject: string, newTrip: INewTrip): Promise<HttpWrapper<ITripDetails>> {
        return (await (await (await HttpWrapper.Init(newTrip))
            .ValidateInput(c => {
                return newTrip.timeZone != null ? null : Results.NotFound('Must supply timezone');
            })
            .Set(TripDbService.fillInMissingData(newTrip))
            .Map(c => TripDbService.createNewDynamoRecord(c, subject))
            .MapAsync(c => TripDbService.addBiteTimes(c)))
            .MapAsync(c => this.tripService.createRecord(c)))
            .Map(d => TripDbService.toTripDetails(d));
    }

    public async updateTrip(subject: string, tripId: string, updateTrip: ITripDetails): Promise<HttpWrapper<ITripDetails>> {
        return (await (await HttpWrapper.Init(updateTrip)
            .ValidateInput(c => {
                return tripId === c.tripId ? null : Results.NotFound(`Cannot change tripId from [${tripId}] to [${c.tripId}]`);
            })
            .MapAsync(() => this.tripService.readRecordWithSortKey(subject, updateTrip.tripId)))
            .Map(c => TripDbService.updateTrip(c, updateTrip))
            .MapAsync(c => this.tripService.updateTripInDynamoDb(c)))
            .Map(c => TripDbService.toTripDetails(c));
    }

    public async patchTrip(subject: string, tripId: string, updateTrip: IUpdateTripDetails): Promise<HttpWrapper<ITripDetails>> {
        return await (await (await (await this.tripService.readRecordWithSortKey(subject, tripId))
            .Map(c => TripDbService.patchTrip(c, updateTrip)))
            .MapAsync(c => this.tripService.updateTripInDynamoDb(c)))
            .Map(c => TripDbService.toTripDetails(c));
    }

    public async deleteTrip(subject: string, tripId: string): Promise<HttpWrapper<ICatchDetails[]>> {
        return await (await (await (await (await this.tripService.readRecordWithSortKey(subject, tripId))
            .Map(c => TripDbService.toTripDetails(c))
            .MapAsync(c => this.tripService.deleteRecord(subject, c.tripId)))
            .MapAsync(c => this.catchService.readAllRecordsForPartition(IdGenerator.generateTripKey(subject, c.TripId))))
            .MapEachAsync<IDynamoDbCatch, IDynamoDbCatch>(c => this.catchService.deleteRecord(IdGenerator.generateTripKey(subject, c.TripId), c.CatchId)))
            .Map(c => c.map(r => CatchDbService.toCatchDetails(r)));
    }

    public async endTrip(subject: string, tripId: string, trip: IEndTripDetails): Promise<HttpWrapper<ITripDetails>> {
        return (await this.catchService.readAllRecordsForPartition(IdGenerator.generateTripKey(subject, tripId)))
            .MapAsync(async all => {
                return (await (await this.tripService.readRecordWithSortKey(subject, tripId))
                    .Map(c => this.tripService.endTrip(c, trip, all.length))
                    .MapAsync(c => this.tripService.updateTripInDynamoDb(c)))
                    .Map(c => TripDbService.toTripDetails(c));
            })
    }
}
